sale_order_product_recommendation/wizards/sale_order_recommendation.py

```diff
@@ -87,8 +91,9 @@ def _generate_recommendations(self):
         found_dict = {l["product_id"][0]: l for l in found_lines}
         RecomendationLine = self.env["sale.order.recommendation.line"]
         existing_product_ids = []
-        # Add products from sale order lines
-        for line in self.order_id.order_line:
+        # Add products from sale order lines that are active and can be sold
+        for line in self.order_id.order_line.filtered(
+                lambda x: x.product_id.active and x.product_id.sale_ok):
```

Comment:
```suggestion
                lambda x: x.product_id.active and x.product_id.sale_ok):
            if not (line.product_id.active and line.product_id.sale_ok):
                continue
```

This should avoid double-looping.

Reply:
In this loop I have actives and sale_or products, so I do nor need this condition

Reply:
Yes, I mean that you can remove the `.filtered()` and use a normal condition instead, to avoid double-looping.

Other wise Odoo loops once for the filtered and a 2nd time for the rest of execution.